Redesign RSS subsystem

glassez opened the pull request:
This is implementation of #6505.

**DESCRIPTION**

The RSS subsystem consists of two main parts:
1. RSS Session (class `RSS::Session`). It manages RSS feeds and downloading its data (articles).
2. RSS Torrent Auto Downloader (class `RSS::AutoDownloader`). It manages auto-downloading rules and downloads torrent from matched RSS articles.

The components mentioned above are persistent application components. We can access/change its property every time (add/edit/remove feeds, add/edit/remove auto downloading rules etc.). Thus, processing can be either enabled or disabled.
The RSS Items (Feeds and Folders) are identified by its paths. We can access some item via `RSS::Session` interface by its path, or via its parent folder interface (only iterate through all folder items). Items don't keep pointer to its parent (since it is not required for read access, and write access is required much less frequently, we get parent via item path).

The RSS components store its data in JSON format under `%UserDataFolder%/rss` directory. They converts its data from legacy format if new files are still not exist.

The RSS GUI is minimally modified to work with the new subsystem. I tried not to change their logic (except in those places where it seemed very horrible). I also changed the names of some classes/files to a more appropriate/correct.

Added new Options page with RSS settings. Added new icon for it (modified existing rss icon by change its color to fit other options page icons).

The following PR can introduce RSS related WebAPI. If nobody wants to do it soon after this PR is merged, I'll do it myself.

glassez commented:
@magao, @mgziminsky, you are welcome.
I gave a General description above. The changes are very large, so I propose to begin not with the view diffs, but with testing. I hope for your help since I do not use these features for personal reasons so I can miss something. I tried not to change the user-visible logic, with the exception of some places (where I thought it was wrong). So I want to get from you a report what is not working as it should (differs from the existing behavior).

@sledgehammer999 approveFeature
Even though you already approved it in #6505, but it affects the GUI a bit, so I want you to look at that part. 

@evsh, @Chocobo1, @ngosang, Although, as I understand it, you are not interested in RSS, but look here too. At least you can notice any visible problems with the code.

mgziminsky commented:
I'll try to test this out and review it this weekend

evsh commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,374 @@
+
+void AutoDownloader::process()
+{
+    if (m_processingQueue.isEmpty()) return; // was be disabled
```
Wounded comment :)

glassez replied:
Fixed.

evsh commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,374 @@
+
+void AutoDownloader::processJob(const QSharedPointer<ProcessingJob> &job)
+{
+    foreach (const QString &ruleName, m_rules.keys()) {
```
Consider iterators or `std::hash` to avoid unnecessary look-ups. Or, better, forget already about Q containers...

Chocobo1 replied:
>forget already about Q containers...

you mean using the counterparts in std lib right?
is there a catch doing so? if no, I'm all for it.

glassez replied:
I'll just change it with `foreach` by value. Please, no start holy war here!

glassez replied:
No, will not work. After all, I need links to the rules... Will use c++11 `for`.

evsh commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,374 @@
+        if (!rule.feedURLs().contains(job->feedURL)) continue;
+        if (!rule.matches(job->articleTitle)) continue;
+
+//        rule.setLastMatch(QDateTime::currentDateTime());
```
?

glassez replied:
What?
If you confusing with commented lines here I do a some optimization (see below), but I leave old code commented here to undo my changes if it is not approved.

evsh replied:
Could you leave a comment in the source code, please?

glassez replied:
The comment below isn't enough? Or I should put it here instead?

evsh replied:
I believe the comment would be placed better right where the question arises. Although, in my code I usually use `#ifdef` in such cases:
```c++
#if USE_OPTIMIZED_CODE_FOR_FEATURE_X
 blah-blah optimized
#else
 vanilla blah-blag
#endif
```

evsh commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,374 @@
+
+        // if rule is in ignoring state do nothing with matched torrent
+        if (rule.ignoreDays() > 0) {
+            // Optimization: update lastMatch only for rules which use it for "ignoring" purposes
```
which -> that

glassez replied:
Fixed.

evsh commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,374 @@
+        BitTorrent::Session::instance()->addTorrent(job->torrentURL, params);
+
+        if (BitTorrent::MagnetUri(job->torrentURL).isValid()) {
+            auto feed = Session::instance()->feedByURL(job->feedURL);
```
Consider
```C++
if (auto feed = Session::instance()->feedByURL(job->feedURL)) {
```

evsh commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,374 @@
+
+void AutoDownloader::load()
+{  
+    QFile rulesFile(m_fileStorage->storageDir().absoluteFilePath(RulesFile));
```
Consider renaming `RulesFile` into `RulesFileName`, because now this line looks confusing.

glassez replied:
Fixed.

evsh commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,374 @@
+    }
+    else if (rulesFile.open(QFile::ReadOnly)) {
+        loadRules(rulesFile.readAll());
+        rulesFile.close();
```
Redundant call.

glassez replied:
Ok. Removed.

evsh commented on src/base/settingsstorage.cpp:
```diff
@@ -30,6 +30,7 @@
 #include "settingsstorage.h"
 
 #include <memory>
+#include <QDebug>
```
Such a strange lone change! Where is the code that requires this include?

glassez replied:
I forgot to remove it after some testing.
By the way, if you're interested, I found one unpleasant thing in qsettings (apparently it refers to INI format).

evsh commented on src/base/tristatebool.cpp:
```diff
@@ -28,33 +28,67 @@
-{
-}
+#include <QDebug>
+#include <QJsonValue>
```
Uhm... A simple short class depends on JSON object? Are you sure?

glassez replied:
Remove it here.

evsh commented on src/base/tristatebool.cpp:
```diff
@@ -28,33 +28,67 @@
+        return jsonVal.toBool();
+
+    if (!jsonVal.isNull())
+        qDebug() << Q_FUNC_INFO << "Incorrect value" << jsonVal.toVariant();
```
"Undefined" is not semantically equal to "incorrect" , but you seems to use "Undefined" in this way. All in all, I'm against this operator because there is no real conversion between a JSON value and bool value and it is illogical to pretend the other way.

glassez replied:
But I need to store TriStateBool in JSON and I also need failsafe conversion from QJsonValue.
Moved it from here to another place.

evsh commented on src/base/tristatebool.cpp:
```diff
@@ -28,33 +28,67 @@
-    case False:
-        m_value = value;
+    switch (i) {
+    case -1: return TriStateBool::Undefined; break;
```
What was wrong with constants???

glassez replied:
We have constants now but in another way. The old approach had problems with comparison operators.
What do you think, will it be better to use `boost::optional<bool>` instead?

evsh commented on src/base/utils/misc.cpp:
```diff
@@ -503,9 +503,9 @@ QList<bool> Utils::Misc::boolListfromStringList(const QStringList &l)
-    const QString scheme = QUrl(s).scheme();
-    QRegExp is_url("http[s]?|ftp", Qt::CaseInsensitive);
-    return is_url.exactMatch(scheme);
+    static const QRegExp isURL("http[s]?|ftp", Qt::CaseInsensitive);
```
The variable name matches nor its value nor usage. `isURLScheme`?

Chocobo1 replied:
If you care about performance you might want to use `QRegularExpression` instead of making it `static`.

>The QRegularExpression class introduced in Qt 5 is a big improvement upon QRegExp, in terms of APIs offered, supported pattern syntax and speed of execution.

glassez replied:
I just made it to be `static const` since it is unchangeable inside method and between its calls.
This is a simple change (like changing the case of characters in a variable name). I don't want to be distracted by outside things in this big PR. So if something is wrong with it, I'll just take it as it was.

evsh replied:
I feel a strong temptation to copy this answer and use it next time I get a review like

> For a starters, fix coding style.

Would you find that appropriate, Sir?

glassez replied:
What do you mean? Did I break some coding style here?

evsh replied:
This universal answer caught my attention:

>  I don't want to be distracted by outside things in this big PR. So if something is wrong with it, I'll just take it as it was.

glassez replied:
@evsh, your memory is strange... 
>For a starters, fix coding style.

Don't forget that when you publish your own PR. In addition, the adherence of coding style is not something external, but an integral part of each coding job in the project.
As for this place, I rewrote it (because it is faster than arguing here with you).
>If you are concerned about performance you might want to use QRegularExpression instead of making it static.

@Chocobo1, I leave it as static const. I don't see any reason why it shouldn't be static. What do you think?

Chocobo1 replied:
>I don't see any reason why it shouldn't be static. What do you think?

I'll let you decide, since I can't argue the slowness of `QRegExp` is at construction or `QRegExp.exactMatch()` (hence the suggestion above).

evsh commented on src/gui/optionsdlg.cpp:
```diff
@@ -84,6 +87,7 @@ OptionsDialog::OptionsDialog(QWidget *parent)
     m_ui->tabSelection->item(TAB_CONNECTION)->setIcon(GuiIconProvider::instance()->getIcon("network-wired"));
     m_ui->tabSelection->item(TAB_DOWNLOADS)->setIcon(GuiIconProvider::instance()->getIcon("folder-download"));
     m_ui->tabSelection->item(TAB_SPEED)->setIcon(GuiIconProvider::instance()->getIcon("speedometer", "chronometer"));
+    m_ui->tabSelection->item(TAB_RSS)->setIcon(GuiIconProvider::instance()->getIcon("rss-config"));
```
This is not a FDO correct icon name. The only one that can be use here and be consistent with FDO scheme is `application-rss+xml` AFAIK. So, suggest the following change:
```c++
m_ui->tabSelection->item(TAB_RSS)->setIcon(GuiIconProvider::instance()->getIcon("application-rss+xml", "rss-config"));
```

glassez replied:
We already have `application-rss+xml` icon, but it isn't I need. It has different color with other options dialog icons. What I can do in this case?

evsh replied:
Sorry, I mistyped. The correct code:
```C++
m_ui->tabSelection->item(TAB_RSS)->setIcon(GuiIconProvider::instance()->getIcon("rss-config", "application-rss+xml"));
```
It will try to load icon "rss-config" from the current theme, this icon will not be found and it will try "application-rss+xml" (also from theme). This will be found. When system icon theme is disabled, the first one will be found.

Chocobo1 commented on src/base/asyncfilestorage.cpp:
```diff
@@ -0,0 +1,113 @@
+#include <QSaveFile>
+#include <QTimer>
+
+AsyncFileStorage::AsyncFileStorage(const QDir &storageDir, QFile *lockFile, QObject *parent)
```
>Why don't you say this in the signature? 

please explain

>the class owns lockFile object

what about `m_lockFile->setParent(this);`?

evsh replied:
The standard way in C++ to say that a function takes ownership of an object, passed by pointer, is to take it as `std::unique_ptr<T>`.

glassez replied:
>Or make this constructor private.

It's private.

glassez replied:
>what about using m_lockFile->setParent(this);?

AFAIK, QFile isn't QObject.

Chocobo1 replied:
but it is... [QFile::QFile(QObject *parent)](http://doc.qt.io/qt-5/qfile.html#QFile-2)

evsh replied:
> It's private.

Indeed, pardon me. My next question: why do you need that `create()` function when you throw an exception if returned object is `nullptr`?

glassez replied:
>but it is... QFile::QFile(QObject *parent)

Oops... Apparently I was too sleepy when I wrote these lines of code. But I think in this case it is not so important to do `m_lockFile->setParent(this);` or `delete m_lockFile`.

glassez replied:
>My next question: why do you need that create() function when you throw an exception if returned object is nullptr?

I do it (break program run) at a higher level. It's too simple (utility) class to do this. It should have a wider scope (for example, somewhere we may need to select a different folder, if we can't "block" some one).

evsh replied:
But this is exactly the point why use exceptions! They pass error information to the level where it can be adequately processed, and if this processing is just logging or re-throwing.

glassez replied:
Maybe I'm behind the times (because I'm just hobbyist), but two equivalent approach existed previously: the use of exceptions or the function return codes/results. It was a matter of taste which to use. I never liked exceptions. If you have no objective arguments, then let it remain as it is.

evsh replied:
I, of course, disagree with you :) They are not equivalent and never were: exception handles... well, an exceptional case, clearly separating in the code a regular branch and the case of an error. I hope some time you agree with this vision. In any case, I do not insists on replacing this function with constructor + exception.

glassez replied:
I dare say that you're probably wrong. Once in the professional environment there is no consensus on this matter, it is impossible to objectively recognize one of these approaches is undoubtedly the best.

evsh replied:
As you like.

glassez replied:
@evsh, if I change my mind and I decide to use exceptions here, how should I do this? Do I need to create `AsyncFileStorageError` inherited from `std::runtime_error`?

evsh replied:
@glassez: exactly.

Chocobo1 commented on src/base/tristatebool.h:
```diff
@@ -29,25 +29,32 @@
 
+class QJsonValue;
+
 class TriStateBool
```
wanna ask, is this class serves the same purpose as `boost::optional<bool>`?

ps. `boost::optional` is header only
pps. c++17 has `std::optional`

evsh replied:
or there is `experimental/optional`. Don't know what compilers have this file.

Chocobo1 commented on src/base/rss/rss_article.cpp:
```diff
@@ -0,0 +1,161 @@
+QJsonObject Article::toJsonObject() const
+{
+    return {
+        {QStringLiteral("id"), m_guid},
```
just looked up `QStringLiteral`.
from http://blog.qt.io/blog/2014/06/13/qt-weekly-13-qstringliteral/
>If you use QStringLiteral you should avoid declaring the same literal in multiple places: This furthermore blows up the binary sizes

glassez replied:
Ok. I'll turn them into static constants.

Chocobo1 commented on src/base/asyncfilestorage.cpp:
```diff
@@ -0,0 +1,113 @@
+AsyncFileStorage *AsyncFileStorage::create(const QString &storageFolder, QObject *parent)
+{
+    QDir storageDir(storageFolder);
+    QFile *lockFile = new QFile(storageDir.absoluteFilePath(QStringLiteral("storage.lock")));
```
would you want to use `QLockFile` ?

glassez replied:
See my comment below.

glassez replied:
Fixed.

Chocobo1 commented on src/base/asyncfilestorage.cpp:
```diff
@@ -0,0 +1,113 @@
+void AsyncFileStorage::processPendingJobs()
+{
+    m_jobsMutex.lock();
+    QStringList fileNames = m_jobs.keys();
```
Not sure I understand thoroughly,
I see `mutex.lock()` everywhere, why not:
```c++
QHash<QString, QByteArray> temp;
m_jobsMutex.lock();
m_jobs.swap(temp);
m_jobsMutex.unlock();

// start operations with temp
```
and no more locking after that.

glassez replied:
In addition to the asynchronous files write another optimization of this class is the following: when someone tries to write the same file (of course with amended contents) repeatedly in a time when a thread is busy (writing another file), only its latest version will be writte. In other words, if we add the job in the queue for file we already have an existing job, we replaced it. 
So your proposal is not appropriate here.

thalieht commented on src/base/base.pri:
```diff
@@ -99,14 +101,14 @@ SOURCES += \
+    $$PWD/rss/rss_folder.cpp \
+    $$PWD/rss/rss_session.cpp \
+	$$PWD/rss/rss_autodownloader.cpp \
+	$$PWD/rss/rss_autodownloadrule.cpp \
```
Something's wrong with the spacing.

thalieht commented on src/base/preferences.h:
```diff
@@ -333,18 +333,8 @@ class Preferences: public QObject
-    QStringList getRssFeedsAliases() const;
-    void setRssFeedsAliases(const QStringList &rssAliases);
+    bool isRSSWidgetVisibled() const;
+    void setRSSWidgetVisibled(const bool enabled);
```
I don't think Visibled is a word. Maybe you meant Visible.

thalieht commented on src/base/rss/private/rss_parser.cpp:
```diff
@@ -349,28 +341,28 @@ void Parser::parseAtomArticle(QXmlStreamReader &xml)
-                QString link = ( xml.attributes().isEmpty() ?
-                                     xml.readElementText().trimmed() :
-                                     xml.attributes().value("href").toString() );
+                QString link = (xml.attributes().isEmpty()
```
In `Parser::parseRssArticle` you change `link` to `tmp` but not here.

glassez replied:
Reverted there.

glassez commented on src/base/asyncfilestorage.cpp:
```diff
@@ -0,0 +1,113 @@
+    QDir storageDir(storageFolder);
+    QFile *lockFile = new QFile(storageDir.absoluteFilePath(QStringLiteral("storage.lock")));
+    if (storageDir.exists() || storageDir.mkpath(storageDir.absolutePath())) {
+        if (lockFile->open(QFile::WriteOnly))
```
Maybe I just chose the wrong name? This is not IPC LockFile. Here I used the same approach as in Bittorrent::Session before. It's just a way to block certain folder from deletion while program run.

evsh replied:
Aha, this makes better sense then. But it will not work in the UNIX world. You have to set immutable flag on the file.
Example: http://stackoverflow.com/questions/32488017/how-do-i-chattr-i-an-open-file-descriptor-in-c

glassez replied:
We can delete folder with opened for writing files on UNIX? Damn, how to live with it? And what happens with this file from the point of view of the writing process?

evsh replied:
> Damn, how to live with it?

Very conveniently. This is the same type of possibility as overwriting running executable.

> And what happens with this file from the point of view of the writing process?

Nothing. It continues to exist until closing.

glassez replied:
>Nothing. It continues to exist until closing.

I don't understand it now. I mean, I can write to a file which no longer exists on disk? Okay, have to fill in the gaps in my knowledge of Unix. 
In any case, I'm incompetent to do anything more advanced with this, so I'll leave it as is. Maybe later someone will come up with a reliable way to solve this...

glassez commented on src/base/asyncfilestorage.cpp:
```diff
@@ -0,0 +1,113 @@
+    while (!fileNames.isEmpty()) {
+        foreach (const QString &fileName, fileNames) {
+            m_jobsMutex.lock();
+            const QByteArray data = m_jobs.take(fileName);
```
AsyncFileStorage emit `failed` signal.

evsh replied:
But I meant that you removed the object from `m_jobs`. and it was not serialised.

glassez replied:
The write fails only in exceptional situations. I don't think we need this data after. The caller decides what to do in case of failure.

evsh replied:
My question is as follows: it is very easy to provide here the strong guarantee, and why don't you do that? Use cases might include network attached drives.

glassez replied:
What do you mean exactly? Should we re-add job into queue if it fails? How many times do we have to try to repeat it in case of failure?

evsh replied:
Strong guarantee means that state of the object does not change in case of error. Thus, try write an object, `throw` or `emit` when it fails or remove the object from collection when writing succeeded.  Then, At the upper level, you may ask user shall we try again. Because maybe he will put cable back in the socket and the problem is solved?

glassez replied:
But this class was designed as non interactive. It's mostly for writing data with short lifetime. Instructed and forgotten. Write error in this case is just a coincidence (which is very unlikely), or serious problems with the hardware, the OS (or brain) of the user. In the first case, it is sufficient to ignore this error and continue working (well, maybe to repeat the task if it has not yet been reset). In the second case, actually, the normal operation of the program impossible, but it is beyond the scope of this PR.

evsh replied:
Assuming your explanation, why signal serialisation error then? Aren't you want to do this to let your client to do something about the error?

glassez replied:
Really, I don't know what to do in this situation. Signal to display a message in the log, while this is the only way. 
All the rest of the application assumes that the data/configuration folders is OK, or just trying to create them every time they want to write the file.
Have you any ideas (not violating the desired behavior in this class)?

evsh replied:
I would forward the decision to the user (human). I.e. I would, as I wrote earlier, implement the strong guarantee, and then ask user whether to try again or not. And remove the key that leads to error in the latter case.

glassez replied:
That's a very good theoretical idea ;) But from a practical point of view, it causes a lot of questions seem to require a separate task to solve them.

glassez replied:
Strong guarantee. 
Perhaps the most simple in itself. If we leave the job in the queue in case of failure, we simply overwrite it as you normally would, if updated data become available.

User interaction. 
We can't implement this for non-interactive component. We can only delegate it in the interactive layer, but how? Now we do not have a suitable notifications subsystem so that it will require to spoil the interfaces of the affected classes by ugly inappropriate signals/methods.

One more question.
Let's say we have the ability to communicate with the user. What can he do in this case? As I said, it's either some kind of temporary glitch that does not require any response, except, perhaps, fix this event in the log and attempts to repeat writing some time ago (agree, I have to implement it), or there is a serious malfunction of the whole system (what to do in this case?).

In General, if you don't have a ready solution (not abstract advice), let's leave this for a separate topic. It doesn't break current application behavior.

glassez commented on src/base/rss/private/rss_parser.cpp:
```diff
@@ -29,15 +29,16 @@
  * Contact : chris@qbittorrent.org
  */
 
+#include "rss_parser.h"
```
To separate an additional prefix from the base name of the file (which usually matches the name of the class). Should I remind you why we use these additional prefixes for files in subdirectories of the project? Since we can have the same file names in different subfolders, it can lead to conflict when you build using qmake (because it builds everything in one folder).

evsh replied:
Thanks, the reminder was needed.

glassez commented on src/base/rss/private/rss_parser.cpp:
```diff
@@ -206,11 +207,23 @@ namespace
+
+Parser::Parser(QString lastBuildDate)
+{
+    qRegisterMetaType<ParsingResult>();
```
I just didn't want to do it in some external class. Admittedly, it's not the best solution. Best to fix it. Tell me please the best way. Maybe some static initializer?

evsh replied:
Yes, I would go with a static initialiser.

glassez replied:
Done.

glassez commented on src/base/rss/private/rss_parser.cpp:
```diff
@@ -206,11 +207,23 @@ namespace
-// read and create items from a rss document
 void Parser::parse(const QByteArray &feedData)
 {
+    QMetaObject::invokeMethod(this, "parse_impl", Qt::QueuedConnection
```
Why for the main thread? Queued slots are called in thread of its object. In case the parser is in another thread it will be called there. We use parsers in another thread (`Session::workingThread()`).

evsh replied:
Ah, Qt's thread magic... Good.

glassez commented on src/base/asyncfilestorage.h:
```diff
@@ -24,35 +24,38 @@
 public:
-    explicit RssSettingsDlg(QWidget *parent = 0);
-    ~RssSettingsDlg();
+    static AsyncFileStorage *create(const QString &storageFolder, QObject *parent = nullptr);
```
We can't create a valid object, if we do not get access to the folder.

evsh replied:
So `thow`then. Why not?

glassez commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,374 @@
+    connect(m_ioThread, &QThread::finished, m_fileStorage, &AsyncFileStorage::deleteLater);
+    connect(m_fileStorage, &AsyncFileStorage::failed, [](const QString &fileName, const QString &errorString)
+    {
+        Logger::instance()->addMessage(QString("Couldn't save RSS AutoDownloader data in %1. Error: %2")
```
A question of the project level. We mix it everywhere.

glassez commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,374 @@
+    m_processingTimer->setSingleShot(true);
+    connect(m_processingTimer, &QTimer::timeout, this, &AutoDownloader::process);
+
+    if (m_processingEnabled) {
```
Ok. Done.

glassez commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,374 @@
+    return m_rules.values();
+}
+
+void AutoDownloader::setRule(const AutoDownloadRule &rule)
```
In the General case, Yes. But how can you be here? To replace this with `addRule()` - `changeRule()`? This will lead to a lot of checks (e.g., adding a rule with a used name, changing the rule, which is still missing).

evsh replied:
Well... Qt names this `insert()`, STL `insert_or_assign()`. IMO, both names are fine.

glassez replied:
When it comes to collections, I agree `insert` is correct name. But AutoDownloader is not a collection of rules, it just contains a collection of rules, so `insert` here is incorrect to. But since I can't find a more appropriate name, I change it on `insertRule`, since it is more like you. Although, maybe `useRule` will be better?

sledgehammer999 commented:
>@sledgehammer999 approveFeature
Even though you already approved it in #6505, but it affects the GUI a bit, so I want you to look at that part.

I already approved as you said. Plus as we talked recently I don't really care about the rss subsystem. So you review and merge this without me being much involved.
I see that you touch parts of session and the main gui. I'll look into them sometime tomorrow and confirm that I am ok with them.

glassez commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,374 @@
+    if (m_rules.contains(newRuleName)) return false;
+
+    m_rules.insert(newRuleName, m_rules.take(ruleName));
+    emit ruleRenamed(newRuleName, ruleName);
```
Done.

glassez commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,374 @@
+void AutoDownloader::removeRule(const QString &ruleName)
+{
+    if (m_rules.remove(ruleName) > 0) {
+        emit ruleRemoved(ruleName);
```
Done.

glassez commented:
@evsh, @Chocobo1, @thalieht, I responded to your comments and corrected code.

thalieht commented on src/gui/rss/rsswidget.h:
```diff
@@ -24,9 +26,8 @@
  */
+
 #ifndef __RSS_IMP_H__
 #define __RSS_IMP_H__
```
You forgot to change these names along the filename change.
Also i noticed you use `#pragma once` in half the header files but include guards in the other half.

glassez replied:
Will be fixed.

>Also i noticed you use `#pragma once` in half the header files but include guards in the other half.

Almost all the files under *base/rss* have been heavily rewritten (or rather, they were created from scratch using the old code) and I decided to use `#pragma once` there. My goal was not to change it everywhere.

mgziminsky commented on src/base/rss/rss_feed.cpp:
```diff
@@ -0,0 +1,418 @@
+    , m_url(url)
+{
+    m_dataFileName = QString("%1.json").arg(
+                QString(QCryptographicHash::hash(m_url.toUtf8(), QCryptographicHash::Sha1).toHex()));
```
Why not give these a more human friendly name? Maybe `m_url.replace(QRegulareExpression("\\W+"), "_")`

glassez replied:
Why do you need human friendly name for these files? It's not for user editing (and even reading in most cases). Are you sure you really want it?

mgziminsky replied:
If you don't intend for a person to be able to read and modify them, don't store them in json. Personally, there have been several times where I have wanted to view and edit this data, so I think you should store them as you are now, but make the filenames more friendly.

glassez replied:
Ok. I'll do it.

mgziminsky commented on src/base/rss/rss_feed.cpp:
```diff
@@ -0,0 +1,418 @@
+{
+    // Download the RSS Feed icon
+    // XXX: This works for most sites but it is not perfect
+    auto iconUrl = QString("http://%1/favicon.ico").arg(QUrl(m_url).host());
```
This should be using the same scheme as the feed url for the icon. This is an existing bug that I had fixed in one of my PRs. https://github.com/qbittorrent/qBittorrent/pull/6368/files#diff-5890b862f225705b8036dc3fbfc71f79L321

mgziminsky commented on src/base/rss/rss_folder.cpp:
```diff
@@ -0,0 +1,132 @@
+    foreach (Item *item, items()) {
+        int n = news.size();
+        news << item->articles();
+        // NOTE: Do we really need this sorting?
```
Answer: Yes

RSS feeds should always be ordered by date

glassez replied:
I mean, the sorting is presentation layer job, isn't it? Someone may need other sorting...

mgziminsky replied:
True, if you were to just do the sorting in the UI that would be fine also. How it is now though, the UI doesn't let you sort manually. I also noticed they are in the reverse order while I was testing.

So either enable sorting in the current UI, with newest at the top being the default, or leave the sorting here and reverse the order.

glassez replied:
Sorting is still here. Or this comment so annoying to you that I should remove it? 
As for the order... You have several times accused me for things I didn't do. This is the existing logic. If you want me to fix something in passing, make a reservation to that effect, why should I do so (why the old logic is incorrect).

mgziminsky replied:
First off, I apologize if my comments seemed accusatory, that wasn't my intention. I'm just trying to describe the expected behavior, and where I found that behavior doesn't align while testing. 

I'm aware that the sorting is still happening in the code below, I just wanted to describe why the sorting was desired. I agree with your response that performing the sorting during presentation would actually be preferred. That isn't currently what is happening though, so it is either necessary to leave the sorting below, or you could remove it from here and add it to the UI.

Because this diff shows everything as added code, it is extremely hard to tell which parts are new, and which parts are unchanged from before. That being said, the article list in the UI is definitely reversed in this version compared to the main version, so something must have changed somewhere.

glassez replied:
The old code had an inconsistent structure, and I may have misunderstood something. How it needs to be sorted from the user's point of view (in the UI)?

glassez replied:
One more question: how articles are usually sorted in feed xml (retrieved from server)?

mgziminsky replied:
The user expects to see the newest articles at the top, which is also typically the order they appear in the feed.

glassez replied:
Hell, one small typo can become a major headache. Fixed it. Thank you for testing!

mgziminsky commented on src/base/rss/rss_item.cpp:
```diff
@@ -0,0 +1,118 @@
+    //        return result;
+
+    int index = 0;
+    while ((index = path.indexOf('\\', index)) >= 0) {
```
You should be using `Item::PathSeparator` here

mgziminsky commented on src/base/rss/rss_item.cpp:
```diff
@@ -0,0 +1,118 @@
+
+bool Item::isValidPath(const QString &path)
+{
+    static const QRegularExpression re(R"(^([^\\]|[^\\]([^\\]|\\(?=[^\\]))*[^\\])$)");
```
This seems a bit lenient. How about just restricting it to word characters (\w), dash (-), and dot (.)?

Also, should use `Item::PathSeparator`

So: `QString(R"(\A\w+(%1\w+)*\z)").arg(Item::PathSeparator)`

glassez replied:
We should not limit this greatly. After all, we assume URL as the last Feed path item.

mgziminsky replied:
Right, I forgot about that. In that case, you can still greatly simplify the regex to my suggestion, just replace both `\w` with `[^%1]`

glassez replied:
My regex mean "have no separator at start and at end and have no two or more separators together". Do your regex mean the same?

mgziminsky replied:
Effectively, yes. Mine is "Start with 1 or more non-separators, followed by 0 or more sets of 1 separator then 1 or more non-separators."

glassez commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,379 @@
+
+    auto feed = Session::instance()->feedByURL(job->feedURL);
+    if (feed) {
+        if (auto article = feed->articleByGUID(job->articleGUID))
```
Well, it's a matter of taste... Maybe I'll rework it in places where it will not break readability.

glassez commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -0,0 +1,379 @@
+        if (rule.ignoreDays() > 0) {
+            // Optimization: update lastMatch only for rules that use it for "ignoring" purposes
+            const QDateTime lastMatch = rule.lastMatch();
+            rule.setLastMatch(QDateTime::currentDateTime());
```
>Setting it here means rules with ignoreDays could be ignored forever.

Please re-read the code. You're wrong.
>I prefer you set it after the if, because I find it useful to know when a rule last matched, even when I'm not using ignored days.

...as it was before.
Ok. I'll produce two compile time branches (as @evsh suggested) and use old behavior by default.

glassez replied:
Or even otherwise. We can skip the immediate rules file overwrite if a rule does not use ignoreDays. It will be updated later. What do you think?

mgziminsky replied:
> You're wrong.

I'm pretty certain I'm not. I'll walk you through what your code is doing now.
Say I have a rule that matches once per day, and `ignoreDays = 2`. It should match every other day, but with yours, the following will happen.

Day 0 - Should download:
 ` rule.lastMatch() = null`
  `rule.setLastMatch(0)`
  `if (null.isValid())` is false. Match success and download. **OK**

Day 1 - Should ignore:
  `rule.lastMatch() = 0`
  `rule.setLastMatch(1)`
  `if (0.isValid())` is true
  `if (1 < (0 + 2))` is true. Skip. **OK**

Day 2 - Should Download:
  `rule.lastMatch() = 1`
  `rule.setLastMatch(2)`
  `if (1.isValid())` is true
  `if (2 < (1 + 2))` is true. Skip. **WRONG**. This will continue forever

mgziminsky replied:
Actually, I also just noticed that you are using `currentDateTime` for the match date, which is also incorrect. You should be using the date from the article.

glassez replied:
Sorry, I wrote the last comment without seeing the code (I remember it was supposed to work, so I thought that this is the case). You're right. I fixed it. Also I defer all kinds of optimization yet...
>Actually, I also just noticed that you are using currentDateTime for the match date, which is also incorrect. You should be using the date from the article.

This is legacy behavior.
Fixed now.

zywo commented:
qbt crashes when removing a download rule (just after pressing OK)


[gdb.txt](https://github.com/qbittorrent/qBittorrent/files/924119/gdb.txt)

mgziminsky commented on src/base/rss/rss_session.cpp:
```diff
@@ -0,0 +1,489 @@
+    m_instance = this;
+
+    m_confFileStorage = AsyncFileStorage::create(
+                Utils::Fs::expandPathAbs(specialFolderLocation(SpecialFolder::Data) + ConfFolder));
```
I think you should use the `SpecialFolder::Config` for this and make `ConfFolder = "rss"` and `ItemsFile = "feeds.json"`. I also think you should place the download rules in the same location. Only the articles should in the data folder.

Speaking of the config storage overall, I think you should change some of the folder names as well. Here is what I think makes the most sense:
`CONFIG_PATH/rss/feeds.json`
`CONFIG_PATH/rss/download_rules.json`
`DATA_PATH/rss/articles/NORMALIZED_FEED_NAME.json`

glassez replied:
Ok. This makes sense.
The one thing I don't like (in general) download rules file contains not only configuration data but also some "state" data that, moreover, are updated frequently (I mean lastMatch field).

mgziminsky replied:
True, but why is that a problem? We already store various other state data in the main settings that are potentially changed frequently, such as UI dimensions. 

It's also potentially useful to treat the lastMatch as config data as well. For instance, it could be useful to be able to clear and/or modifying that value when also using ignore days to manipulate the next time a match will occur.

glassez replied:
I always thought this is incorrect, and I have repeatedly offered to separate settings and states. 
But here you are right (I thought about this too), lastMatch can be considered as settings as well.

glassez commented:
>qbt crashes when removing a download rule (just after pressing OK)

@zywo, Fixed now! Thanks.

glassez commented:
@mgziminsky, @Chocobo1, @evsh, all your comments are addressed (I hope I didn't forget something).
Rebased ot top of current master.
Waiting for your new comments...
@sledgehammer999, you promised to check some part of this too.

Chocobo1 commented on src/base/asyncfilestorage.h:
```diff
@@ -1,6 +1,6 @@
- * Bittorrent Client using Qt4 and libtorrent.
- * Copyright (C) 2010  Christophe Dumez
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2015  Vladimir Golovnev <glassez@yandex.ru>
```
2017.

glassez replied:
Fixed.

sledgehammer999 commented on src/base/rss/rss_session.h:
```diff
@@ -0,0 +1,153 @@
+
+    class Session: public QObject
+    {
+        Q_OBJECT
```
Don't you need `Q_DISABLE_COPY(Session)` too?

sledgehammer999 commented on src/base/rss/rss_session.h:
```diff
@@ -0,0 +1,153 @@
+    {
+        Q_OBJECT
+
+        friend class ::Application;
```
Isn't it a good idea to keep things the similar? All of our other singletons use the initInstance()/freeInstance()/instance() triplet. Unless you want to solve a technical problem...

glassez replied:
This is my new approach to make singletons initialization more clear. The first created object become global instance (similar approach is used in QApplication). Since all our singletons must be initialized by Application there is no problem. It allow to initialize singletons using its constructors. It will allow to clearly initialize inherited singletons (like GuiIconProvider; we just call `new GuiIconProvider`and then use `IconProvider::instance()` in base code).

sledgehammer999 commented on src/base/preferences.cpp:
```diff
@@ -1241,32 +1241,32 @@ void Preferences::setRssHSplitterSizes(const QByteArray &sizes)
 QStringList Preferences::getRssOpenFolders() const
 {
-    return value("Rss/open_folders").toStringList();
+    return value("GUI/RSSWidget/OpenedFolders").toStringList();
```
Why not keep the old names? Does the data format change here?
(same to the others)

glassez replied:
I had some conflicts with my new config options (because of some strange/specific/buggy QSettings behavior I found). I prefer rename old options to proper names instead of trying to fit the new names.
This will break only RSS GUI settings and make it invisible again. But it would further indicate to the user that there is now a new (revised) RSS subsystem.

sledgehammer999 replied:
I didn't notice that these are for GUI settings. So it is ok to break them IMO.

sledgehammer999 commented on src/base/asyncfilestorage.h:
```diff
@@ -24,35 +24,38 @@
+    QFile *m_lockFile;
+    QHash<QString, QByteArray> m_jobs;
+    QMutex m_jobsMutex;
+    bool m_isIdle = true;
```
should be initialized in the constructor

glassez replied:
It will be initialized in every constructor that does not initialize it explicitly.
This new syntax ensures that the field will be initialized and we can not worry that we will forget to do it in some constructor. I would strongly not recommend to ignore this initializer form and/or to prohibit it.

sledgehammer999 replied:
So is this even better than doing it the initializer list?
I think you have only one constructor here.
Just so you know, this design decision will result in recompilations of stuff that include this header in case we want to change the value of the variable. (I am talking about local testing. Releasing requires clean builds anyway).

glassez replied:
>Just so you know, this design decision will result in recompilations of stuff that include this header in case we want to change the value of the variable. 

But this is only the default value! It does not change practically never. In any case, much less than the other members of the class.

sledgehammer999 replied:
Ok leave as it is. I am not going to argue for this particular variable. Under normal circumstances this value won't ever change, so further discussing is moot.

glassez commented:
>This seems very similar to functionality from DownloadManager. What is the advantage of AsyncFileStorage over it?

I don't understand what similarity you found in these classes... They do different things. The first loads the data from the network, the second saves the data to disk.
The main AsyncFileStorage goals:
1. Storing the data asynchronously (e.g. in separate thread)
2. Do not save outdated file data if updated is available (when updated file data is ready but the outdated one is still in queue we don't append new job, we replace the old one)

And Yes, I plan to use it globally later.

glassez commented:
@evsh, I don't know whether it is your changes on the portable settings result or not, but I noticed that `specialFolderLocation(SpecialFolder::Config)` is not the same folder where application settings are stored. What do you say?

glassez commented:
@magao, are you able to test this PR too (since you are interested in RSS, IIRC)?

evsh commented:
>  I noticed that `specialFolderLocation(SpecialFolder::Config)` is not the same folder where application settings are stored. What do you say?

Explain, please. Don't understand what do you mean. Specifically, what do you expect and what do you get from it?

glassez commented:
>Explain, please. Don't understand what do you mean. Specifically, what do you expect and what do you get from it?

I expect it is the folder where qBittorrent.ini are stored (I can't provide you exact path since I'm far from my computer) but it really is `config` subfolder under qBittorrent data folder (where we have BT_backup etc.). If you need exact path, I can provide it later.

evsh commented:
With 3.3.12 installation, `qBittorrent.ini` is located in `%USERPROFILE%/AppData/Roaming/qBittorrent`. `Private::DefaultProfile::configLocation()` on Windows returns `Private::DefaultProfile::dataLocation()` plus `/config/`, which must be wrong. Additionally, Windows sub-path ignores configuration name. See #6651, please.

sledgehammer999 commented on src/base/preferences.cpp:
```diff
@@ -1410,64 +1410,14 @@ void Preferences::setTransHeaderState(const QByteArray &state)
+bool Preferences::isRSSWidgetVisible() const
 {
-    return value("Preferences/RSS/RSSEnabled", false).toBool();
+    return value("GUI/RSSWidget/Visible", false).toBool();
```
Comment about consistency: IIRC the rest of the settings(in settingstorage too) use "Enabled" to denote the visibility of gui element. I am just mentioning it. Do as you please here. It isn't a hard requirement for me.

glassez replied:
Fixed.

sledgehammer999 commented:
>I don't understand what similarity you found in these classes... They do different things. The first loads the data from the network, the second saves the data to disk.
The main AsyncFileStorage goals:
1. Storing the data asynchronously (e.g. in separate thread)
2. Do not save outdated file data if updated is available (when updated file data is ready but the outdated one is still in queue we don't append new job, we replace the old one)

And Yes, I plan to use it globally later.

First of all I didn't look at the usage of the class. Just its implementation. So it wasn't clear to do from where the data came from. I assumed you save something that you download. Under this light I don't understand why you don't tell the downloadhandler to save to file.
Now I see that you use the new class to store JSON(?) objects. Since I don't want to look at the whole RSS subsystem, can you describe very shortly what is the workflow? (so I can get an idea)

glassez commented:
When the RSS feeds are updated, the AutDownoloader starts to process the Articles, applying to them the rules. Because this is done in the main thread, we don't process them all at once, and use deferred processing while the application is idle. Each time the rule matched some article we download its torrent, mark article as read and update rule matchDate. Then we store file with rules data and file with affected feed data. (This is current behavior). Since disk io is slow it's better to do it in non gui thread. The steps described above can be repeated frequently within a short period of time, so we can get updated data for some file not yet written.
This is my first approach to optimize it.

sledgehammer999 commented:
>When the RSS feeds are updated, the AutDownoloader starts to process the Articles, applying to them the rules. Because this is done in the main thread, we don't process them all at once, and use deferred processing while the application is idle. Each time the rule matched some article we download its torrent, mark article as read and update rule matchDate. Then we store file with rules data and file with affected feed data. (This is current behavior). Since disk io is slow it's better to do it in non gui thread. The steps described above can be repeated frequently within a short period of time, so we can get updated data for some file not yet written.
This is my first approach to optimize it.

OK. So you make AsyncFileStorage run on its own thread to do the saving...
Just a thought for the **future**(I haven't seen the whole interaction so I might be talking rubbish here): It would be interesting to use AsyncFileStorage to savefiles to disk in the `DownloadHandler` class.(which takes an argument when to saveToFile or keep in memory)

glassez commented:
>It would be interesting to use AsyncFileStorage to savefiles to disk in the DownloadHandler class.(which takes an argument when to saveToFile or keep in memory)

It makes sense to use it everywhere we need non interactive file saving. I want to do this in the future, because there are some issues associated with non-interactive save data (like what to do in case of failure, etc.). We discussed this with @evsh here, but while this is only theoretical arguments and suggestions. This topic deserves special attention.

glassez commented:
I want to simplify AsyncDataStorage by removing jobs rewriting. I want optimize the RSS classes to perform delayed data saving (like in SettingsStorage). But I want to do it in separate PR to not delay this PR merging. There are some other jobs need to be done on top of it.
@sledgehammer999?

sledgehammer999 commented:
@glassez The only think you didn't answer is if you need to put `Q_DISABLE_COPY(Session)` too.
PS: @evsh your request is blocking merge. If you're done, approve so @glassez can merge.

evsh commented:
I don't understand yet what is planned to do with the `AsyncFileStorage` class. Right now it looks self-inconsistent: its interface gives some error-handling hint, but implementation does not provide this opportunity. And there are two little suggestions (I'll write them right away).

glassez commented:
>The only think you didn't answer is if you need to put Q_DISABLE_COPY(Session) too.

I'll do it.

evsh commented on src/base/asyncfilestorage.cpp:
```diff
@@ -0,0 +1,113 @@
+
+AsyncFileStorage::~AsyncFileStorage()
+{
+    m_lockFile->close();
```
I don't like that this resource is created not in the constructor, but is freed in the destructor. AFAIK,  local throw/catch in 
`AsyncFileStorage::create()` but with all the invariant set up code in the constructor would be better.

glassez replied:
Ok. I'll redo it using constructor+exceptions.

evsh commented on src/base/asyncfilestorage.cpp:
```diff
@@ -0,0 +1,113 @@
+    QDir storageDir(storageFolder);
+    if (storageDir.exists() || storageDir.mkpath(storageDir.absolutePath())) {
+        QFile *lockFile = new QFile(storageDir.absoluteFilePath(QStringLiteral("storage.lock")));
+        if (lockFile->open(QFile::WriteOnly))
```
If you are not going to implement the lock for UNIX filesystems, please leave a `// TODO:` comment or anything like that.

glassez replied:
Ok. I'll do it.
One question on the future. Is there a portable way to do this is for all UNIX systems? Or it depends on the file system?

evsh commented on src/base/asyncfilestorage.cpp:
```diff
@@ -0,0 +1,113 @@
+    if (m_isIdle) {
+        m_isIdle = false;
+        // The function will be run in the thread of receiver (this AsyncFileStorage object).
+        QTimer::singleShot(0, this, &AsyncFileStorage::processPendingJobs);
```
Maybe an implementation via a critical section would be a cleaner one? Just asking, did try to implement.

glassez replied:
>I want to simplify AsyncDataStorage by removing jobs rewriting. I want optimize the RSS classes to perform delayed data saving

So this "concurrency" code will be unneeded. 
Think of it as a temporary solution. I'll change it, perhaps even along with this PR. Either directly after it. As I wrote, I don't want to delay it, because the basic functionality is ready and it is required for other tasks that need to be on top of it (I mean so long-awaited web interface).

sledgehammer999 commented on src/base/asyncfilestorage.cpp:
```diff
@@ -0,0 +1,116 @@
+        throw AsyncFileStorageError(
+                QString("Could not create directory '%1'.").arg(m_storageDir.absolutePath()));
+
+    // TODO: This folder locking approach does not work for UNIX systems. Implement it.
```
Hmm, interesting. I didn't know that.
So just a heads up: Any solution applied here should be applied to the "BT_backup" lockfile too.

glassez replied:
Of course.

sledgehammer999 commented:
One stupid question/reminder, because I don't intend to read the whole RSS code. Did you remember to cleanup the files written by `AsyncFileStorage` when they're not needed anymore?

glassez commented:
When user remove some feed it removes associated file. All other rss files are used permanently. If you mean lock file, it is deleted in destructor.

glassez commented:
>Selecting an item while viewing the "Unread" entries doesn't mark the article as read

It's not bug, it's a feature (at least it should be). The reason is when we select some article it immediately become read and it should be removed from unread list so we don't have time to read it actually. Although now I see that I chose the wrong approach. Why don't we mark articles as read after the user has selected a different article?

glassez commented:
@mgziminsky, can you provide me some non-private feed urls for testing?

mgziminsky commented:
@glassez here are a couple feeds.
https://www.nyaa.se/?page=rss
https://www.nyaa.se/?page=rss&user=64513

glassez commented:
>The alert message about download rules being disabled shouldn't resize with the rest of the dialog

I don't understand what do you mean...

glassez commented:
@mgziminsky, I've updated the code. The issue with application crash when downloading torrent from RSS should be fixed now. Also I've changed "read article" behavior, as I suggested above.

mgziminsky commented:
If you disable download rules in the settings, then open the download rules dialog, there is a message at the top informing you that rules are disabled globally. If you resise the dialog, the section at the top with the message scales up as well, filling the dialog with empty space.

I'll test again when I get off work tonight.

glassez commented:
>If you resise the dialog, the section at the top with the message scales up as well, filling the dialog with empty space.

OK. I realized what it was. But I'm not so familiar with GUI design, so I can't fix it. I can't understand why it works differently here and in the main window.

LordNyriox commented:
@Chocobo1: ???

mgziminsky commented:
To fix the UI issue, you just need to set the vertical size policy of hsplitter to expanding. One more question about the rules dialog though. Did you disable the import/export buttons because you are planning to re-implement that feature? If not, I think it would be better to just remove them entirely. 

Other than that, I'm not seeing any more crashes and things seem to be working.

glassez commented:
>To fix the UI issue, you just need to set the vertical size policy of hsplitter to expanding.

Ok. Fixed now.
>One more question about the rules dialog though. Did you disable the import/export buttons because you are planning to re-implement that feature? If not, I think it would be better to just remove them entirely.

I'll remove it before next release unless we have some other ideas about it.

glassez commented:
@mgziminsky I've pushed one more commit. Please review it.
The reason is we should add new articles in chronological order (I mean the adding process only - the article list is in reverse order as expected) to prevent wrong ignore behavior (when rule ignores more recent article).

Matth7878 commented:
Not that I use import/export feature but I think a lot of people would want it back. Besides isn't it bad to loose feature which are perceived as basic?
Once again I really don't care about this feature and I'm really glad (and grateful) to see RSS being overhauled but it should be important to know why people are using it. IMHO it's to automatically download from rules or to start downloading​ from a webui.

glassez commented:
One more thing I want to fix here (or in the following PR) is "paranoid" data saving policy of RSS classes (it was be in old code and I still leave it as is).

As for imports/exports... Previously, these functions were used to transfer the rules between instances of the qBittorrent (e.g. on different computer), because it is stored inside the other settings, and it was impossible to transfer them directly. Now rules are stored in a separate file and we can just copy it. So import/export in the old sense is no longer needed. Maybe this need for exchange of rules with the third-party apps...

glassez commented:
@mgziminsky, shouldn't we mark all articles as read after AutoDownloader has it processed (regardless of whether matched any rule or not)? I think it's logical, it read them all!

mgziminsky commented:
Good catch on the article load order.

As for marking them all read from the AutoDownloader, I think we should definitely not do that. When I set up rules, it's because there are specific entries I expect to show up in a feed. In those cases, it's desired for the rule to "read" the article on my behalf because I will know when they were there because they will show up in my torrent downloads. For the others, I still want to manually inspect what new entries are showing up in my feeds so that I can decide if I want to create new rules or not.

Along those lines, one feature I've been considering adding is "ignore rules" which would filter articles I know I know that I don't care about out of the feed. Then you would have the ability to only ever see "new types" of articles.

glassez commented:
>Along those lines, one feature I've been considering adding is "ignore rules" which would filter articles I know I know that I don't care about out of the feed. Then you would have the ability to only ever see "new types" of articles.

I think it's separate job (out of this PR), isn't it?

mgziminsky commented:
yes, definitely a separate PR, I'm not asking you to add it to this one. Once this is merged, I plan to continue where I left off with my currently open PRs, so I'll probably add it myself.

glassez commented:
@mgziminsky, I've pushed one more commit (probably the last in this PR). Please test it. If it's OK, i'll merge this PR.

glassez commented:
@evsh, Cmake build is broken now (after this PR was meged). Please fix it.

glassez commented:
@mgziminsky, @evsh, @thalieht, @zywo, @Chocobo1, thank you for your cooperation.

glassez commented:
@mgziminsky, @magao, apparently, I will oversee RSS subsystem for a while until we get a handle on "hot" PRs. Please make revision of all RSS PRs in terms of their relevance, and give your comments about them.
